### PR TITLE
Feat: Add Redis watcher

### PR DESCRIPTION
**Pull request by MilesChou**
Call facade `Redis::enableEvents()` before using this watcher.

### CODE REVIEW

**Review by brettmc — APPROVED** [review]

**Comment by MilesChou** [comment]
@brettmc How can I improve this PR.

**Comment by brettmc** [comment]
> How can I improve this PR.

I think we looked at this in last week's SIG meeting, and this one was good to merge, @ChrisLightfootWild ?

**Comment by ChrisLightfootWild** [comment]
> > How can I improve this PR.
> 
> I think we looked at this in last week's SIG meeting, and this one was good to merge, @ChrisLightfootWild ?

Apologies for my very delayed response @MilesChou - I did promise to take a look at both of your PRs this week and I have unfortunately only just gotten around to it. Sorry! :heart:

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/Serializer.php [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+        foreach (self::SERIALIZATION_SUBSETS as $subset) {
+            if (preg_match($subset['regex'], $command)) {
+                $paramsToSerializeNum = $subset['args'];
+
+                break;
+            }
+        }
+
+        // Serialize the allowed number of arguments
+        $paramsWillToSerialize = ($paramsToSerializeNum >= 0) ? array_slice($params, 0, $paramsToSerializeNum) : $params;
```
[the remark this answers is not in the document]
Comment (MilesChou) [reply src/Instrumentation/Laravel/src/Watchers/RedisCommand/Serializer.php]: LGTM

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+            ->setSpanKind(SpanKind::KIND_CLIENT)
+            ->setStartTimestamp($this->calculateQueryStartTime($nowInNs, $event->time))
+            ->startSpan();
+
+        // See https://opentelemetry.io/docs/specs/semconv/database/redis/
+        $attributes = [
+            TraceAttributes::DB_SYSTEM => TraceAttributeValues::DB_SYSTEM_REDIS,
+            TraceAttributes::DB_NAME => $this->fetchDbIndex($event->connection),
+            TraceAttributes::DB_OPERATION => $operationName,
+            TraceAttributes::DB_QUERY_TEXT => Serializer::serializeCommand($event->command, $event->parameters),
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php:62]: This appears to be unavailable in semconv 1.24, which is the target for a 1.0 release (and the last supported version for php 8.0).

Not sure if there is an alternative which might be better suited here?

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+                /** @psalm-suppress PossiblyUndefinedMethod */
+                return $connection->client()->getConnection()->getParameters()->database;
+            }
+        } catch (Throwable $e) {
+        }
+
+        return null;
+    }
+
+    private function fetchDbHost(Connection $connection): string
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php:91]: This throws exceptions still and should probably also return null if a value cannot be resolved correctly.

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+            if ($connection instanceof PhpRedisConnection) {
+                return $connection->client()->getDbNum();
+            } elseif ($connection instanceof PredisConnection) {
+                /** @psalm-suppress PossiblyUndefinedMethod */
+                return $connection->client()->getConnection()->getParameters()->database;
+            }
+        } catch (Throwable $e) {
+        }
+
+        return null;
```
Comment (ChrisLightfootWild) [inline src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php:88]:
```suggestion
        } catch (Throwable $e) {
            return null;
        }
```

Comment (MilesChou) [reply src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php]: Fix in 633a56e

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php (hunk 2 above)
Comment (MilesChou) [reply src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php]: Use DB_STATEMENT for compatibility with semconv 1.24.

Ref: https://github.com/open-telemetry/opentelemetry-specification/blob/v1.19.0/specification/trace/semantic_conventions/database.md#redis

**Review by ChrisLightfootWild — APPROVED** [review]

File: src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+use Illuminate\Redis\Connections\PhpRedisConnection;
+use Illuminate\Redis\Connections\PredisConnection;
+use Illuminate\Redis\Events\CommandExecuted;
+use OpenTelemetry\API\Instrumentation\CachedInstrumentation;
+use OpenTelemetry\API\Trace\SpanKind;
+use OpenTelemetry\Contrib\Instrumentation\Laravel\Watchers\Watcher;
+use OpenTelemetry\SemConv\TraceAttributes;
+use OpenTelemetry\SemConv\TraceAttributeValues;
+use RangeException;
+use RuntimeException;
```
[the remark this answers is not in the document]
Comment (MilesChou) [reply src/Instrumentation/Laravel/src/Watchers/RedisCommand/RedisCommandWatcher.php]: Removed 😅

**Pull request merged**